hadml/models/cgan/multihadron_event_gan.py

```diff
@@ -18,52 +18,92 @@ def __init__(
         discriminator: torch.nn.Module,
         optimizer_generator: Optimizer,
         optimizer_discriminator: Optimizer,
-        noise_dim: int
+        noise_dim: int,
+        loss_type: str,
+        r1_reg: float,
+        target_gumbel_temp: float = 0.3
     ):
         super().__init__()
         self.save_hyperparameters(ignore=["generator", "discriminator"])
         self.generator = generator
         self.discriminator = discriminator
+        self.train_gen_loss = MeanMetric()
+        self.train_disc_loss = MeanMetric()
+        self.val_gen_loss = MeanMetric()
+        self.val_disc_loss = MeanMetric()
+        self.current_gumbel_temp = 1.0
 
     def forward(self, clusters):
-        generator_input = conditional_cat(clusters, self._generate_noise(len(clusters)))
-        generated_hadrons = self.hparams.generator(generator_input)
+        noise = self._generate_noise(*clusters.size()[:2])
+        generated_hadrons = self.generator(noise.to(clusters.device), clusters)
+        hadron_kins_dim = self.generator.hadron_kins_dim
+        generated_hadrons[:, :, hadron_kins_dim:] = torch.nn.functional.gumbel_softmax(
```

Comment:
Can you make the `hard` argument configurable through the config file ([link](https://pytorch.org/docs/stable/generated/torch.nn.functional.gumbel_softmax.html#torch-nn-functional-gumbel-softmax))? It would be worthwhile to test whether it improves the performance (I'm expecting that it will)